pkg/core/region.go

```diff
@@ -1826,37 +1827,75 @@ func (r *RegionsInfo) ScanRegions(startKey, endKey []byte, limit int) []*RegionI
 // BatchScanRegions scans regions in given key pairs, returns at most `limit` regions.
 // limit <= 0 means no limit.
 // The given key pairs should be non-overlapping.
-func (r *RegionsInfo) BatchScanRegions(keyRanges *KeyRanges, limit int) []*RegionInfo {
-	r.t.RLock()
-	defer r.t.RUnlock()
-
+func (r *RegionsInfo) BatchScanRegions(keyRanges *KeyRanges, limit int, outputMustContainAllKeyRange bool) ([]*RegionInfo, error) {
+	keyRanges.Merge()
 	krs := keyRanges.Ranges()
 	res := make([]*RegionInfo, 0, len(krs))
-	var lastRegion *RegionInfo
+
+	r.t.RLock()
+	defer r.t.RUnlock()
 	for _, keyRange := range krs {
 		if limit > 0 && len(res) >= limit {
-			return res
+			res = res[:limit]
+			return res, nil
 		}
-		if lastRegion != nil {
-			if lastRegion.Contains(keyRange.EndKey) {
-				continue
-			} else if lastRegion.Contains(keyRange.StartKey) {
-				keyRange.StartKey = lastRegion.GetEndKey()
+
+		regions, err := scanRegion(r.tree, keyRange, limit)
+		if err != nil {
+			log.Warn("scan regions failed", zap.Bool("outputMustContainAllKeyRange", outputMustContainAllKeyRange), zap.Error(err))
+			if outputMustContainAllKeyRange {
+				return nil, err
 			}
 		}
-		r.tree.scanRange(keyRange.StartKey, func(region *RegionInfo) bool {
-			if len(keyRange.EndKey) > 0 && bytes.Compare(region.GetStartKey(), keyRange.EndKey) >= 0 {
-				return false
-			}
-			if limit > 0 && len(res) >= limit {
-				return false
-			}
-			lastRegion = region
-			res = append(res, region)
-			return true
-		})
+		if len(res) > 0 && len(regions) > 0 && res[len(res)-1].meta.Id == regions[0].meta.Id {
+			// skip the region that has been scanned
+			regions = regions[1:]
+		}
+		res = append(res, regions...)
 	}
-	return res
+	return res, nil
+}
+
+func scanRegion(regionTree *regionTree, keyRange *KeyRange, limit int) ([]*RegionInfo, error) {
+	var (
+		res        []*RegionInfo
+		lastRegion = &RegionInfo{
+			meta: &metapb.Region{EndKey: keyRange.StartKey},
+		}
+		exceedLimit = func() bool { return limit > 0 && len(res) >= limit }
+		err         error
+	)
+	regionTree.scanRange(keyRange.StartKey, func(region *RegionInfo) bool {
+		if len(keyRange.EndKey) > 0 && len(region.GetStartKey()) > 0 &&
+			bytes.Compare(region.GetStartKey(), keyRange.EndKey) >= 0 {
+			return false
+		}
+		if exceedLimit() {
+			return false
+		}
+		if len(lastRegion.GetEndKey()) > 0 && len(region.GetStartKey()) > 0 &&
+			bytes.Compare(region.GetStartKey(), lastRegion.GetEndKey()) > 0 {
+			err = multierr.Append(err, errors.Errorf(
```

Comment:
if we pass `outputMustContainAllKeyRange` into this function, we can return `false` to `regionTree.scanRange` to fail fast. Also I think a large error message may exceed gRPC message size limit.

Reply:
> if we pass outputMustContainAllKeyRange into this function, we can return false to regionTree.scanRange to fail fast.

Yeah, it's a trade-off. I chose not to pass in any extra parameters to avoid making this function too complicated, but that does mean it can't fail fast. Waiting to hear what other reviewers think.

> Also it can avoid a large error message exceeding gRPC message size limit.

This function won't cause any issues with gRPC transmission, so there's no problem. Regions with issues will be dropped at the upper level.